### ISSUE TITLE
perf: move compute-heavy note commitment tree updates to a non-blocking worker thread in zebra-state

### ISSUE THREAD

**Issue by dconnolly** [ZcashFoundation/zebra#4297]
## Motivation

@conradoplg did some benchmarking on where `zebrad` does the most work during sync, and it looks like updating the note commitment Merkle trees is expensive. Currently this work is done as part of the main blocking thread.

### Designs

One thing we can do is move this currently blocking, compute-heavy work to a non-blocking async thread, to optimize the need to swap in and out of threads, and unblock other work in the main thread while the expensive tree updates are happening on non-blocking threads.

This would/could be done for all the note commitment trees: Orchard, Sapling, and Sprout.

## Related Work

Closed but incomplete:
- #2188
- #2202

**Comment by teor2345** [comment]
This isn't required for the `lightwalletd` work, so I've set it to low priority.

**Comment by teor2345** [comment]
We tried to fix this in PR https://github.com/ZcashFoundation/zebra/pull/4199.

It worked, but it didn't help performance much, and the extra threading caused some fragile tests to intermittently fail.
(We can fix the tests, we just decided we didn't have time right now.)
